src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -12,26 +12,20 @@ namespace System.Net.Http
     internal class WinHttpResponseStream : Stream
     {
         private volatile bool _disposed = false;
-        private SafeWinHttpHandle _sessionHandle = null;
-        private SafeWinHttpHandle _connectHandle = null;
         private SafeWinHttpHandle _requestHandle = null;
 
-        internal WinHttpResponseStream(
-            SafeWinHttpHandle sessionHandle,
-            SafeWinHttpHandle connectHandle,
-            SafeWinHttpHandle requestHandle)
+        internal WinHttpResponseStream(SafeWinHttpHandle requestHandle)
         {
-            // While we only use the requestHandle to do actual reads of the response body,
-            // we need to keep the parent handles (connection, session) alive as well.
             bool ignore = false;
-            sessionHandle.DangerousAddRef(ref ignore);
-            connectHandle.DangerousAddRef(ref ignore);
             requestHandle.DangerousAddRef(ref ignore);
-            _sessionHandle = sessionHandle;
-            _connectHandle = connectHandle;
             _requestHandle = requestHandle;
         }
 
+        ~WinHttpResponseStream()
+        {
+            Dispose(false);
```

Comment:
The _requestHandle is originally created by WinHttpHandler during process of sending an HTTP request with SendAsync(). After the HTTP request has returned a response from the server, a WinHttpResponseStream object is created and the _requestHandle is passed into its constructor. WinHttpResponseStream does DangerousAddRef in the constructor.  Meanwhile the WinHttpHandler SendAsync() finishes and does Dispose() on the _requestHandle.  At this point, the _requestHandle is still alive since the response stream has done the DangerousAddRef.  When the response stream is closed (Dispose() or finalizer), DangerousRelease is done and the _requestHandle can go away.

So, to answer your question:

>or b) some code elsewhere called Dispose on the handle while its ref count was zero. The latter shouldn't happen, as these handles aren't expose out to arbitrary code, right?
The _requestHandle is not exposed out to arbitrary code.  But it is Disposed via the ending of the SendAsync() method in the WinHttpHandler.

The only way to avoid doing the AddRef/Release and needing a finalizer in WinHttpResponseStream is to avoid doing the Dispose() of the _requestHandle in the WinHttpHandler at the conclusion of the SendAsync() processing.  I'll need to look at whether that hand-off of the _requestHandle can be done without doing the extra AddRef/Release.